llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp

```diff
@@ -885,53 +885,51 @@ template <typename Ty = llvm::Value> Ty *getVal(llvm::Metadata *M) {
   return nullptr;
 }
 
-/// Return the MDNode that has the SLM size attribute.
-static llvm::MDNode *getSLMSizeMDNode(llvm::Function *F) {
-  llvm::NamedMDNode *Nodes =
+static inline llvm::Metadata *getMD(llvm::Value *V) {
+  return llvm::ValueAsMetadata::get(V);
+}
+
+/// Updates genx.kernels metadata attribute \p MD for the given function \p F.
+/// The value of the attribute is updated only if the new value \p NewVal is
+/// bigger than what is already stored in the attribute.
+static void updateGenXMDNodes(llvm::Function *F, genx::KernelMDOp MD,
+                              uint64_t NewVal) {
+  llvm::NamedMDNode *GenXKernelMD =
       F->getParent()->getNamedMetadata(GENX_KERNEL_METADATA);
-  assert(Nodes && "invalid genx.kernels metadata");
-  for (auto Node : Nodes->operands()) {
-    if (Node->getNumOperands() >= 4 && getVal(Node->getOperand(0)) == F)
-      return Node;
-  }
-  // if F is not a kernel, keep looking into its callers
-  while (!F->use_empty()) {
-    auto CI = cast<CallInst>(F->use_begin()->getUser());
-    auto UF = CI->getParent()->getParent();
-    if (auto Node = getSLMSizeMDNode(UF))
-      return Node;
+  assert(GenXKernelMD && "invalid genx.kernels metadata");
+  for (auto Node : GenXKernelMD->operands()) {
+    if (Node->getNumOperands() > MD &&
+        getVal(Node->getOperand(genx::KernelMDOp::FunctionRef)) == F) {
+
+      llvm::Value *Old = getVal(Node->getOperand(MD));
+      assert((Old && isa<llvm::ConstantInt>(Old)) &&
+             "integer constant expected");
+      uint64_t OldVal = cast<llvm::ConstantInt>(Old)->getZExtValue();
+      if (OldVal < NewVal) {
+        llvm::Value *New = llvm::ConstantInt::get(Old->getType(), NewVal);
+        Node->replaceOperandWith(MD, getMD(New));
+      }
+    }
   }
-  return nullptr;
-}
 
-static inline llvm::Metadata *getMD(llvm::Value *V) {
-  return llvm::ValueAsMetadata::get(V);
+  // Update all callers of 'F' as well.
+  for (auto It = F->use_begin(); It != F->use_end(); It++) {
```

Comment:
I think we need to protect going twice through the same function even if there are no cycles in a call graph, because it is not really a tree. This will protect from cycles as well as a side effect.

Reply:
Fixed that in e39c9ef6a9abd9d2b94a84c9d32929554da3f841